pkg/cluster/util.go

```diff
@@ -0,0 +1,225 @@
+package cluster
+
+import (
+	"archive/zip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/k3s/pkg/daemons/config"
+)
+
+const (
+	backupDest                 = "/tmp/rke1"
+	compressedExtension        = "zip"
+	stateExtenstion            = "rkestate"
+	caCertName                 = "kube-ca"
+	requestHeaderCACertName    = "kube-apiserver-requestheader-ca"
+	etcdClientCACertName       = "kube-etcd-client-ca"
+	serviceAccountTokenKeyName = "kube-service-account-token"
+	certPEM                    = "certificatePEM"
+	keyPEM                     = "keyPEM"
+
+	certType = iota
+	keyType
+)
+
+func unzip(src, dest string) error {
+	r, err := zip.OpenReader(src)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Close(); err != nil {
+			panic(err)
+		}
+	}()
+
+	os.MkdirAll(dest, 0700)
+
+	for _, f := range r.File {
+		err := extract(f, dest)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extract(f *zip.File, dest string) error {
+	rc, err := f.Open()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := rc.Close(); err != nil {
+			panic(err)
+		}
+	}()
+
+	path := filepath.Join(dest, f.Name)
+
+	if !strings.HasPrefix(path, filepath.Clean(dest)+string(os.PathSeparator)) {
+		return fmt.Errorf("illegal file path: %s", path)
+	}
+
+	if f.FileInfo().IsDir() {
+		os.MkdirAll(path, f.Mode())
+	} else {
+		os.MkdirAll(filepath.Dir(path), f.Mode())
+		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := f.Close(); err != nil {
+				panic(err)
+			}
+		}()
+
+		_, err = io.Copy(f, rc)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func isCompressed(filename string) bool {
+	return strings.HasSuffix(filename, fmt.Sprintf(".%s", compressedExtension))
```

Comment:
This can be simplified.

```suggestion
	return strings.HasSuffix(filename, "." + compressedExtension)
```